man/ibcacheedit.8
.TH IBCACHEEDIT 8 "Apr 15, 2010" "OpenIB" "OpenIB Diagnostics"

.SH NAME
ibcacheedit \- edit an ibnetdiscover cache

.SH SYNOPSIS
.B ibcacheedit
[\-\-switchguid BEFOREGUID:AFTERGUID] [\-\-caguid BEFORE:AFTER]
[\-\-sysimgguid BEFOREGUID:AFTERGUID] [\-\-portguid NODEGUID:BEFOREGUID:AFTERGUID]
[\-h(elp)] <orig.cache> <new.cache>

.SH DESCRIPTION
.PP
ibcacheedit allows users to edit an ibnetdiscover cache created through the
\fB\-\-cache\fR option in
.B ibnetdiscover(8).

.SH OPTIONS

.PP
.TP
\fB\-\-switchguid\fR BEFOREGUID:AFTERGUID
Specify a switchguid that should be changed.  The before and after guid
should be separated by a colon.  On switches, port guids are identical
to the switch guid, so port guids will be adjusted as well on switches.
.TP
\fB\-\-caguid\fR BEFOREGUID:AFTERGUID
Specify a caguid that should be changed.  The before and after guid
should be separated by a colon.
.TP
\fB\-\-sysimgguid\fR BEFOREGUID:AFTERGUID
Specify a sysimgguid that should be changed.  The before and after guid
should be spearated by a colon.
.TP
\fB\-\-portguid\fR NODEGUID:BEFOREGUID:AFTERGUID
Specify a portguid that should be changed.  The nodeguid of the port
(e.g. switchguid or caguid) should be specified first, followed by a
colon, the before port guid, another colon, then the after port guid.
On switches, port guids are identical to the switch guid, so the
switch guid will be adjusted as well on switches.

.SH COMMON OPTIONS

Most OpenIB diagnostics take the following common flags. The exact list of
supported flags per utility can be found in the usage message and can be shown
using the util_name -h syntax.

# Debugging flags
.PP
\-h      show the usage message
.PP
\-V      show the version info.

.SH AUTHORS
.TP
Albert Chu
.RI < chu11@llnl.gov >


ChangeLog
commit c3cc42fffb2d1266cb2c3389a5fa3a8fd4f3f9e9
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:52:23 2009 -0500

    Spin 1.2.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit bfdca7d4392aaee9d8253d306378b036b8a60874
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:52:22 2009 -0500

    Remove immediate work requests
    
    The aren't supported/defined in iWARP.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 7b7460e47ad82ea81101230642c43ee10c6ad517
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:18:12 2009 -0500

    Handle 0B IO better.
    
    Allow num_sge == 0 to indicate a 0B IO.
    
    Also, for read, if length is 0 then create a valid 0B read request
    regardless of the user's specified lkey/rkey.  This avoids 0 stag
    connection terminations.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit e901d7d2ab0e9429118765e98e69a3992b91c584
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:00:54 2009 -0500

    Reset the T3 wq when the qp moves to RESET.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0f71e2dc0e1a40f4954f30f14deae965818c3c4e
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:10:57 2009 -0500

    Spin release 1.2.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 76d2a6347ab753d26237dc94c95da8d0c6c0c251
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:07:44 2009 -0500

    Support for doorbell overflow avoidance.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ad77c21cafde669c878032fc0f46635408568498
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:07:20 2009 -0500

    Support sq_sig_all QP attribute.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f6ea9eb207fa4b68f3a9c6f016bc8a262edb1808
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:58:03 2009 -0500

    Spin release version 1.2.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0ec532c2b6b92be6adf58849c9e3d215a50d306d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:52:51 2009 -0500

    Support new device ids: 0x35, 0x36.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f685c8fe7e77e64614d825e563dd9f02a0b1ae16
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:46 2008 -0500

    Spin release 1.2.2
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ebd19d077d331270805aefa9ca00d3661561ba87
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:27 2008 -0500

    Relax the required firmware level.
    
    Turns out that libcxgb3 can really run on either 6.0 or 5.0
    firmware.  So I'm relaxing the requirement...
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 6f7485feb244d8571fcab2292ef92c97bea48df0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue May 13 09:09:35 2008 -0500

    Spin release 1.2.1.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 309916b23ba384a0bd297e7e51b6b89bff357a78
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:45:01 2008 -0500

    Use local stag == 1 to silently drop the p2p read response.
    
    The current code fails if there are read completions -and-
    the p2p read response in the cq when its flushed.
    
    This fix uses the fact that the RTR read is to target stag 1,
    which is not a valid stag.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 67d519c4a9e28e54a9e436404156c9ae7eb60314
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:25:41 2008 -0500

    Handle wrapping when flushing the software sq.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a3dcea751a94ad5e0a9cad6c55a839a9848f327c
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Spin 1.2.0 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit cc2ef741842883233a77e18d65a828c711e581a3
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit 92322b93f48fa98b4fadb736ae49dafda237272a
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Apr 24 14:23:50 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit a8fb6221af4d914f39c1e71f9331f6d428d698c2
Author: Steve Wise (Open Grid Computing) <swise@hosting.openfabrics.org>
Date:   Tue Apr 1 15:30:11 2008 -0700

    Spin 1.1.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit db41ebc10f4cc4c2fe7826589636ab7d4b804218
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 26 08:27:12 2008 -0600

    Provide an empty stub for iwch_query_qp.
    
    This prevents crashes when calling ibv_query_qp due to an expected
    library provided function.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ec4f7f1c6b8beceb9001519bc4e2ee89fbb2e6d0
Author: Dotan Barak <dotanb@dev.mellanox.co.il>
Date:   Tue Feb 26 08:23:36 2008 -0600

    Remove duplicated line.
    
    Signed-off-by: Dotan Barak <dotanb@dev.mellanox.co.il>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 8218482d03898f859e1fa272649fafc64f91f05f
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Feb 12 15:16:03 2008 -0600

    Spin release 1.1.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 1bc6930d8884646ef73db71a95cc32c96c20ba32
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 12 14:46:10 2008 -0600

    libcxgb3: zeroing of wc_flags
    

src/ibcacheedit.c
/*
 * Copyright (c) 2010 Lawrence Livermore National Lab.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <inttypes.h>

#include <infiniband/mad.h>
#include <infiniband/ibnetdisc.h>

#include "ibdiag_common.h"

uint64_t switchguid_before = 0;
uint64_t switchguid_after = 0;
int switchguid_flag = 0;

uint64_t caguid_before = 0;
uint64_t caguid_after = 0;
int caguid_flag = 0;

uint64_t sysimgguid_before = 0;
uint64_t sysimgguid_after = 0;
int sysimgguid_flag = 0;

uint64_t portguid_nodeguid = 0;
uint64_t portguid_before = 0;
uint64_t portguid_after = 0;
int portguid_flag = 0;

struct guids {
	uint64_t searchguid;
	int searchguid_found;
	uint64_t before;
	uint64_t after;
	int found;
};

static int parse_beforeafter(char *arg, uint64_t *before, uint64_t *after)
{
	char *ptr;
	char *before_str;
	char *after_str;

	ptr = strchr(optarg, ':');
	if (!ptr || !(*(ptr + 1))) {
		fprintf(stderr, "invalid input '%s'\n", arg);
		return -1;
	}
	(*ptr) = '\0';
	before_str = arg;
	after_str = ptr + 1;

	(*before) = strtoull(before_str, 0, 0);
	(*after) = strtoull(after_str, 0, 0);
	return 0;
}

static int parse_guidbeforeafter(char *arg,
				 uint64_t *guid,
				 uint64_t *before,
				 uint64_t *after)
{
	char *ptr1;
	char *ptr2;
	char *guid_str;
	char *before_str;
	char *after_str;

	ptr1 = strchr(optarg, ':');
	if (!ptr1 || !(*(ptr1 + 1))) {
		fprintf(stderr, "invalid input '%s'\n", arg);
		return -1;
	}
	guid_str = arg;
	before_str = ptr1 + 1;

	ptr2 = strchr(before_str, ':');
	if (!ptr2 || !(*(ptr2 + 1))) {
		fprintf(stderr, "invalid input '%s'\n", arg);
		return -1;
	}
	(*ptr1) = '\0';
	(*ptr2) = '\0';
	after_str = ptr2 + 1;

	(*guid) = strtoull(guid_str, 0, 0);
	(*before) = strtoull(before_str, 0, 0);
	(*after) = strtoull(after_str, 0, 0);
	return 0;
}

static int process_opt(void *context, int ch, char *optarg)
{
	switch (ch) {
	case 1:
		if (parse_beforeafter(optarg,
				      &switchguid_before,
				      &switchguid_after) < 0)
			return -1;
		switchguid_flag++;
		break;
	case 2:
		if (parse_beforeafter(optarg,
				      &caguid_before,
				      &caguid_after) < 0)
			return -1;
		caguid_flag++;
		break;
	case 3:
		if (parse_beforeafter(optarg,
				      &sysimgguid_before,
				      &sysimgguid_after) < 0)
			return -1;
		sysimgguid_flag++;
		break;
	case 4:
		if (parse_guidbeforeafter(optarg,
					  &portguid_nodeguid,
					  &portguid_before,
					  &portguid_after) < 0)
			return -1;
		portguid_flag++;
		break;
	default:
		return -1;
	}

	return 0;
}

static void update_switchportguids(ibnd_node_t *node, uint64_t guid)
{
	ibnd_port_t *port;
	int p;

	for (p = 0; p <= node->numports; p++) {
		port = node->ports[p];
		if (port)
			port->guid = node->guid;
	}
}

static void replace_node_guid(ibnd_node_t *node, void *user_data)
{
	struct guids *guids;

	guids = (struct guids *)user_data;

	if (node->guid == guids->before) {

		node->guid = guids->after;

		/* port guids are identical to switch guids on
		 * switches, so update port guids too
		 */
		if (node->type == IB_NODE_SWITCH)
			update_switchportguids(node, guids->after);

		guids->found++;
	}
}

static void replace_sysimgguid(ibnd_node_t *node, void *user_data)
{
	struct guids *guids;
	uint64_t sysimgguid;

	guids = (struct guids *)user_data;

	sysimgguid = mad_get_field64(node->info, 0, IB_NODE_SYSTEM_GUID_F);
	if (sysimgguid == guids->before) {
		mad_set_field64(node->info, 0, IB_NODE_SYSTEM_GUID_F,
				guids->after);
		guids->found++;
	}
}

static void replace_portguid(ibnd_node_t *node, void *user_data)
{
	struct guids *guids;

	guids = (struct guids *)user_data;

	if (node->guid != guids->searchguid)
		return;

	guids->searchguid_found++;

	if (node->type == IB_NODE_SWITCH) {
		/* port guids are identical to switch guids on
		 * switches, so update switch guid too
		 */
		if (node->guid == guids->before) {
			node->guid = guids->after;
			update_switchportguids(node, guids->after);
			guids->found++;
		}
	}
	else {
		ibnd_port_t *port;
		int p;

		for (p = 1; p <= node->numports; p++) {
			port = node->ports[p];
			if (port
			    && port->guid == guids->before) {
				port->guid = guids->after;
				guids->found++;
				break;
			}
		}
	}
}

int main(int argc, char **argv)
{
	ibnd_fabric_t *fabric = NULL;
	char *orig_cache_file = NULL;
	char *new_cache_file = NULL;
	struct guids guids;

	const struct ibdiag_opt opts[] = {
		{"switchguid", 1, 1, "BEFOREGUID:AFTERGUID",
		 "Specify before and after switchguid to edit"},
		{"caguid", 2, 1, "BEFOREGUID:AFTERGUID",
		 "Specify before and after caguid to edit"},
		{"sysimgguid", 3, 1, "BEFOREGUID:AFTERGUID",
		 "Specify before and after sysimgguid to edit"},
		{"portguid", 4, 1, "NODEGUID:BEFOREGUID:AFTERGUID",
		 "Specify before and after port guid to edit"},
		{0}
	};
	char *usage_args = "<orig.cache> <new.cache>";

	ibdiag_process_opts(argc, argv, NULL, "edCPDLGtsv",
			    opts, process_opt, usage_args,
			    NULL);

	argc -= optind;
	argv += optind;

	orig_cache_file = argv[0];
	new_cache_file = argv[1];

	if (!orig_cache_file)
		IBERROR("original cache file not specified");

	if (!new_cache_file)
		IBERROR("new cache file not specified");

	if ((fabric = ibnd_load_fabric(orig_cache_file, 0)) == NULL)
		IBERROR("loading original cached fabric failed");

	if (switchguid_flag) {
		guids.before = switchguid_before;
		guids.after = switchguid_after;
		guids.found = 0;
		ibnd_iter_nodes_type(fabric,
				     replace_node_guid,
				     IB_NODE_SWITCH,
				     &guids);

		if (!guids.found)
			IBERROR("switchguid = %" PRIx64 " not found",
				switchguid_before);
	}

	if (caguid_flag) {
		guids.before = caguid_before;
		guids.after = caguid_after;
		guids.found = 0;
		ibnd_iter_nodes_type(fabric,
				     replace_node_guid,
				     IB_NODE_CA,
				     &guids);

		if (!guids.found)
			IBERROR("caguid = %" PRIx64 " not found",
				caguid_before);
	}

	if (sysimgguid_flag) {
		guids.before = sysimgguid_before;
		guids.after = sysimgguid_after;
		guids.found = 0;
		ibnd_iter_nodes(fabric,
				replace_sysimgguid,
				&guids);

		if (!guids.found)
			IBERROR("sysimgguid = %" PRIx64 " not found",
				sysimgguid_before);
	}

	if (portguid_flag) {
		guids.searchguid = portguid_nodeguid;
		guids.searchguid_found = 0;
		guids.before = portguid_before;
		guids.after = portguid_after;
		guids.found = 0;
		ibnd_iter_nodes(fabric,
				replace_portguid,
				&guids);

		if (!guids.searchguid_found)
			IBERROR("nodeguid = %" PRIx64 " not found",
				portguid_nodeguid);

		if (!guids.found)
			IBERROR("portguid = %" PRIx64 " not found",
				portguid_before);
	}

	if (ibnd_cache_fabric(fabric, new_cache_file, 0) < 0)
		IBERROR("caching new cache data failed");

	ibnd_destroy_fabric(fabric);
	exit(0);
}
